pvlib/modelchain.py

```diff
@@ -1410,10 +1680,37 @@ def run_model_from_effective_irradiance(self, data=None):
         pvlib.modelchain.ModelChain.run_model_from
         pvlib.modelchain.ModelChain.run_model_from_poa
         """
-
+        self._check_length(data)
+        _validate_weather_indices(data)
         self._assign_weather(data)
         self._assign_total_irrad(data)
-        self.effective_irradiance = data['effective_irradiance']
+        self.results.effective_irradiance = _tuple_from_dfs(
+            data, 'effective_irradiance')
         self._run_from_effective_irrad(data)
 
         return self
+
+
+def _validate_weather_indices(data):
+    indexes = map(lambda df: df.index, data)
+    if not all(itertools.starmap(pd.Index.equals,
+                                 itertools.combinations(indexes, 2))):
+        raise ValueError("Weather DataFrames must have same index.")
```

Comment:
Might want to make this error message more generic:
```suggestion
        raise ValueError("Input DataFrames must have same index.")
```